pkg/volume/scaleio/sio_util.go

```diff
@@ -0,0 +1,210 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scaleio
+
+import (
+	"errors"
+	"fmt"
+	"path"
+	"strconv"
+
+	"github.com/golang/glog"
+
+	api "k8s.io/kubernetes/pkg/api/v1"
+	meta "k8s.io/kubernetes/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/volume"
+	volutil "k8s.io/kubernetes/pkg/volume/util"
+)
+
+var (
+	confKey = struct {
+		gateway,
+		sslEnabled,
+		secretRef,
+		system,
+		protectionDomain,
+		storagePool,
+		storageMode,
+		sdcRootPath,
+		volumeName,
+		fsType,
+		readOnly,
+		username,
+		password string
+	}{
+		gateway:          "gateway",
+		sslEnabled:       "sslEnabled",
+		secretRef:        "secretRef",
+		system:           "system",
+		protectionDomain: "protectionDomain",
+		storagePool:      "storagePool",
+		storageMode:      "storageMode",
+		sdcRootPath:      "sdcRootPath",
+		volumeName:       "volumeName",
+		fsType:           "fsType",
+		readOnly:         "readOnly",
+		username:         "username",
+		password:         "password",
+	}
+)
+
+// mapScaleIOVolumeSource maps attributes from a ScaleIOVolumeSource to config
+func mapVolumeSource(config map[string]string, source *api.ScaleIOVolumeSource) {
+	config[confKey.gateway] = source.Gateway
+	config[confKey.secretRef] = string(source.SecretRef.Name)
+	config[confKey.system] = source.System
+	config[confKey.volumeName] = source.VolumeName
+
+	//optionals
+	assertConfigDefaults(config)
+}
+
+func validateConfigs(config map[string]string) error {
+	if config[confKey.gateway] == "" {
+		return errors.New("missing gateway")
+	}
+	if config[confKey.secretRef] == "" {
+		return errors.New("missing secret ref")
+	}
+	if config[confKey.system] == "" {
+		return errors.New("missing system")
+	}
+
+	return nil
+}
+
+func assertConfigDefaults(config map[string]string) {
+	b, _ := strconv.ParseBool(config[confKey.sslEnabled])
+	config[confKey.sslEnabled] = strconv.FormatBool(b)
+	config[confKey.protectionDomain] = defaultString(config[confKey.protectionDomain], "default")
+	config[confKey.storagePool] = defaultString(config[confKey.storagePool], "default")
+	config[confKey.storageMode] = defaultString(config[confKey.storageMode], "ThinProvisioned")
+	config[confKey.sdcRootPath] = defaultString(config[confKey.sdcRootPath], "/opt/emc/scaleio/sdc/bin/")
+	config[confKey.fsType] = defaultString(config[confKey.fsType], "ext4")
+	b, _ = strconv.ParseBool(config[confKey.readOnly])
+	config[confKey.readOnly] = strconv.FormatBool(b)
+}
+
+func defaultString(val, defVal string) string {
```

Comment:
defaults go to `pkg/api/v1/defaults.go`. Also add tests in `pkg/api/testing/fuzzer.go`